runtime/memprof.c

```diff
@@ -206,1126 +1258,921 @@ void caml_memprof_update_suspended(bool s) {
+     * sampling at least once, but maybe more than once. */
+    size_t samples = 1 +
+      rand_binom(domain,
+                 Caml_state->memprof_young_trigger - 1 - Caml_state->young_ptr);
+    CAMLassert(encoded_alloc_lens == NULL);
+    maybe_track_block(domain, Val_hp(Caml_state->young_ptr),
+                      samples, wosize, SRC_NORMAL);
+    caml_memprof_set_trigger(Caml_state);
+    caml_reset_young_limit(Caml_state);
+    CAMLreturn0;
   }
 
-  /* We need to call the callbacks for this sampled block. Since each
-     callback can potentially allocate, the sampled block will *not*
-     be the one pointed to by [caml_memprof_young_trigger]. Instead,
-     we remember that we need to sample the next allocated word,
-     call the callback and use as a sample the block which will be
-     allocated right after the callback. */
-
-  CAMLassert(Caml_state->young_ptr < caml_memprof_young_trigger &&
-             caml_memprof_young_trigger <= Caml_state->young_ptr + whsize);
-  trigger_ofs = caml_memprof_young_trigger - Caml_state->young_ptr;
-  alloc_ofs = whsize;
-
-  /* Restore the minor heap in a valid state for calling the callbacks.
-     We should not call the GC before these two instructions. */
+  CAMLassert(Caml_state->young_ptr < Caml_state->memprof_young_trigger &&
+             Caml_state->memprof_young_trigger <=
+               Caml_state->young_ptr + whsize);
+
+  /* In comments in this section, we refer to the individual block
+   * allocations within the combined allocation as
+   * "sub-allocations". */
+
+  /* Trigger offset from the base of the combined allocation. Signed
+   * so it can go negative as we iterate. */
+  intnat trigger_ofs =
+    Caml_state->memprof_young_trigger - Caml_state->young_ptr;
+  /* Sub-allocation offset from the base of the combined
+   * allocation. Signed so we can compare correctly against
+   * trigger_ofs. */
+  intnat alloc_ofs = whsize;
+
+  /* Undo the combined allocation, so that our callbacks run on a
+     consistent heap (and can perform their own allocations). */
   Caml_state->young_ptr += whsize;
-  caml_memprof_set_suspended(1); // This also updates the memprof trigger
 
-  /* Perform the sampling of the block in the set of Comballoc'd
-     blocks, insert them in the entries array, and run the
-     callbacks. */
-  for (alloc_idx = nallocs - 1; alloc_idx >= 0; alloc_idx--) {
-    unsigned alloc_wosz = encoded_alloc_lens == NULL ? wosize :
+  /* set to true if any callback raises an exception */
+  bool cancelled = false;
+  /* how many new entries belong to us? */
+  size_t new_entries = 0;
+
+  /* Suspend profiling while we call allocation callbacks */
+  update_suspended(domain, true);
+
+  /* Work through the sub-allocations, high address to low address,
+   * identifying which ones are sampled and how many times.  For each
+   * sampled sub-allocation, create an entry and run the allocation
+   * callback. */
+  size_t alloc_idx = nallocs;
+  do {
```

Comment:
The logic in this loop is very complicated, because of the need to correctly handle state changing during the loop. Would it be simpler to create all the entries first without calling any callbacks, then call `entries_run_callback_exn` to call them all? (This will additionally run delayed callbacks that happen to be in the list, but that seems good).

Reply:
I like this idea and can't immediately see anything wrong with it. There's a very small change of semantics (if we decide to sample more than one allocation of the comballoc, and one of the callbacks stops the profile, and possibly starts another one), but I think that's a change we can reasonably tolerate. I will try it out.

Reply:
Note: the code here changed quite a bit (and was squash-rebased) between @stedolan's review and @jhjourdan's, so this conversation may be confusing.
Before @stedolan's comment, we had control flow more like that in OCaml 4.14: we iterated through the sub-allocations and for each one decided on sampling, called the callback if appropriate, and then managed the outcome accordingly (including, if lambda changed, changing the state in the loop so that subsequent sub-allocations would be sampled according to the new lambda).
That's no longer the case: now (following @stedolan's suggestion) we create all entries according to the initial lambda, and then run `entries_run_callback_exn`. If a callback stops sampling, then (as @jhjourdan notes) we notice that in `entries_run_callback_exn`. But if a callback stops sampling and starts another profile, then in `caml_memprof_start` we move the entries to an orphans table, which treats offset entries specially: ones which have had their allocation callback called get marked for deallocation; ones which have not had their allocation callback called are deleted.

Reply:
Right. Then perhaps we could document that if profiles are stopped and then restarted in a callback, then we may miss some samples and consider as deallocated even if they aren't. That's not crucial, though.